src/main/kotlin/com/openlattice/data/DataGraphService.kt

```diff
@@ -601,12 +594,51 @@ open class DataGraphService(
             // ensure, that DataEdge src and dst entity types are part of src and dst entity types of AssociationType
             val srcEntityType = edmManager.getEntityTypeByEntitySetId(it.src.entitySetId)
             val dstEntityType = edmManager.getEntityTypeByEntitySetId(it.dst.entitySetId)
-            if (!(associationType.src.contains(srcEntityType.id)
-                            && associationType.dst.contains(dstEntityType.id))) {
+
+            checkAllowedEntityTypesOfAssociation(
+                    associationType,
+                    associationEntitySetId,
+                    setOf(srcEntityType.id),
+                    setOf(it.src.entitySetId),
+                    setOf(dstEntityType.id),
+                    setOf(it.dst.entitySetId))
+        }
+    }
+
+    private fun checkAllowedEntityTypesOfAssociation(
+            edgeAssociationType: AssociationType,
+            edgeEntitySetId: UUID,
+            srcEntityTypes: Collection<UUID>,
+            srcEntitySetIds: Set<UUID>,
+            dstEntityTypes: Collection<UUID>,
+            dstEntitySetIds: Set<UUID>
+    ) {
+        val isSrcMNotAllowed = !edgeAssociationType.src.containsAll(srcEntityTypes)
+        val isDstNotAllowed = !edgeAssociationType.dst.containsAll(dstEntityTypes)
+
+        if (edgeAssociationType.isBidirectional) {
+            val isSrcNotAllowedInDst = !edgeAssociationType.dst.containsAll(srcEntityTypes)
+            val isDstNotAllowedInSrc = !edgeAssociationType.src.containsAll(dstEntityTypes)
```

Comment:
I'm not sure this will do the right thing -- if the src (or dst) entity types that are used are actually divided between the src and dst types on a bidirectional association type, it looks like this will still fail

Reply:
Oh, I see, because it's checking in a batch.. I guess bidirectional cases need to handled individually for each association.